python/grapher.go

```diff
@@ -34,33 +36,64 @@ func NewGraphContext(unit *unit.SourceUnit) *GraphContext {
 // Graphs the Python source unit. If run outside of a Docker container, this assumes that the source unit has already
 // been installed (via pip or `python setup.py install`).
 func (c *GraphContext) Graph() (*graph.Output, error) {
-	if os.Getenv("IN_DOCKER_CONTAINER") != "" {
-		// NOTE: this may cause an error when graphing any source unit that depends
-		// on jedi (or any other dependency of the graph code)
-		requirementFiles, err := filepath.Glob(filepath.Join(c.Unit.Dir, "*requirements*.txt"))
+	programMode := os.Getenv("IN_DOCKER_CONTAINER") == ""
+	tc, err := toolchain.Lookup("sourcegraph.com/sourcegraph/srclib-python")
+	if err != nil {
+		return nil, err
+	}
+	pipBin := "pip"
+	pythonBin := "python"
+
+	if programMode {
+		tempPath, err := getTempPath()
 		if err != nil {
 			return nil, err
 		}
-		runCmdLogError(exec.Command("pip", "install", "-I", c.Unit.Dir))
-		for _, requirementFile := range requirementFiles {
-			err := runCmdStderr(exec.Command("pip", "install", "-r", requirementFile))
-			if err != nil {
-				log.Printf("Error installing dependencies in %s. Trying piecemeal install")
-				if b, err := ioutil.ReadFile(requirementFile); err == nil {
-					for _, req := range strings.Split(string(b), "\n") {
-						runCmdLogError(exec.Command("pip", "install", req))
-					}
-				} else {
-					log.Printf("Could not read %s: %s", requirementFile, err)
-				}
+		envName := fmt.Sprintf("%s-%s-env", getHash(c.Unit.Dir), url.QueryEscape(c.Unit.Name))
+		envDir := filepath.Join(tempPath, envName)
+
+		if _, err := os.Stat(filepath.Join(envDir)); os.IsNotExist(err) {
+			// We don't have virtual env for this SourceUnit, create one.
+			tcVENVBinPath := filepath.Join(tc.Dir, ".env", "bin")
+			cmd := exec.Command(filepath.Join(tcVENVBinPath, "virtualenv"), envDir)
+			if err := runCmdStderr(cmd); err != nil {
+				return nil, err
 			}
 		}
+		// Use binaries from our virutal env.
+		pipBin = filepath.Join(envDir, "bin", "pip")
+		pythonBin = filepath.Join(envDir, "bin", "python")
 	}
-	p, err := getVENVBinPath()
+
+	// NOTE: this may cause an error when graphing any source unit that depends
+	// on jedi (or any other dependency of the graph code)
+	requirementFiles, err := filepath.Glob(filepath.Join(c.Unit.Dir, "*requirements*.txt"))
 	if err != nil {
 		return nil, err
 	}
-	cmd := exec.Command(filepath.Join(p, "python"), "-m", "grapher.graph", "--verbose", "--dir", c.Unit.Dir, "--files")
+	if _, err := os.Stat(filepath.Join(c.Unit.Dir, "setup.py")); !os.IsNotExist(err) {
+		runCmdLogError(exec.Command(pipBin, "install", "-I", c.Unit.Dir))
```

Comment:
Since this code used to be executed only in docker mode, I tried to not change the functionality too much. Previously (in docker mode) it just reported error and continued.

It's kinda fail fast vs analyze at least something. I guess if run locally failing fast makes sense but if you run it on sourcegraph.com you probably want to show at least something to the end user and then maybe report the errors/warning.